src/matvis/core/beams.py

```diff
@@ -0,0 +1,186 @@
+"""Functions for working with beams."""
+
+import numpy as np
+from abc import ABC, abstractmethod
+from pyuvdata import UVBeam
+from typing import Any
+
+
+def _wrangle_beams(
+    beam_idx: np.ndarray | None,
+    beam_list: list[UVBeam],
+    polarized: bool,
+    nant: int,
+    freq: float,
+) -> tuple[list[UVBeam], int, np.ndarray]:
+    """Perform all the operations and checks on the input beams.
+
+    Checks that the beam indices match the number of antennas, pre-interpolates to the
+    given frequency, and checks that the beam type is appropriate for the given
+    polarization
+
+    Parameters
+    ----------
+    beam_idx
+        Index of the beam to use for each antenna.
+    beam_list
+        List of unique beams.
+    polarized
+        Whether to use beam polarization
+    nant
+        Number of antennas
+    freq
+        Frequency to interpolate beam to.
+    """
+    # Get the number of unique beams
+    nbeam = len(beam_list)
+
+    # Check the beam indices
+    if beam_idx is None and nbeam not in (1, nant):
+        raise ValueError(
+            "If number of beams provided is not 1 or nant, beam_idx must be provided."
+        )
+    if beam_idx is not None:
+        assert beam_idx.shape == (nant,), "beam_idx must be length nant"
+        assert all(
+            0 <= i < nbeam for i in beam_idx
+        ), "beam_idx contains indices greater than the number of beams"
+
+    for bm in beam_list:
+        if isinstance(bm, UVBeam) and not bm.future_array_shapes:
+            bm.use_future_array_shapes()
+
+    # make sure we interpolate to the right frequency first.
+    beam_list = [
+        (
+            bm.interp(freq_array=np.array([freq]), new_object=True, run_check=False)
+            if isinstance(bm, UVBeam)
+            else bm
+        )
+        for bm in beam_list
+    ]
+
+    if polarized and any(b.beam_type != "efield" for b in beam_list):
+        raise ValueError("beam type must be efield if using polarized=True")
+    elif not polarized and any(
+        (
+            b.beam_type != "power"
+            or getattr(b, "Npols", 1) > 1
+            or b.polarization_array[0] not in [-5, -6]
+        )
+        for b in beam_list
+    ):
+        raise ValueError(
+            "beam type must be power and have only one pol (either xx or yy) if polarized=False"
+        )
+
+    return beam_list, nbeam, beam_idx
+
+
+class BeamInterpolator(ABC):
```

Comment:
Perhaps, add a note or ToDo as comments / docstrings on how pyuvdata=3.1 will affect this, if any.

Reply:
It shouldn't affect it at all, actually

Reply:
Okay great.